race-buffer/src/writer.rs

```diff
@@ -0,0 +1,252 @@
+            Ok(RaceBuffer {
+                wcurs: 0,
+                owcurs: storage.len(),
+                storage,
+                use_base_2_indexing,
+            })
+        }
+    }
+
+    /// Create new RaceBuffer with properly aligned backing storage
+    #[inline]
+    pub fn new_from_bytes(bytes: &'a mut [u8]) -> Result<RaceBuffer<'a, E>, SizeError> {
+        let (_prefix, buf, _suffix) = Self::align_from_bytes(bytes);
+        buf
+    }
+
+    /// Create new RaceBuffer with properly aligned backing storage, return unused bytes
+    #[inline]
+    pub fn align_from_bytes(
+        bytes: &'a mut [u8],
+    ) -> (
+        &'a mut [u8],
+        Result<RaceBuffer<'a, E>, SizeError>,
+        &'a mut [u8],
+    ) {
+        // Safe because storage is treated as uninit after transmutation
+        let (prefix, storage, suffix) = unsafe { bytes.align_to_mut() };
+        (prefix, RaceBuffer::new(storage), suffix)
+    }
+
+    /// Write single entry to buffer, returning overwritten entry
+    pub fn write(&mut self, entry: E) -> OverwrittenEntry<E> {
+        let mut overwritten = OverwrittenEntry::None;
+        if self.wcurs == self.owcurs {
+            // Reading storage is safe after checking if buffer has looped at least once
+            if self.wcurs >= self.storage.len()
+                && unsafe { self.read_storage(self.wcurs).is_prefix() }
+            {
+                // If overwriting double, increment overwrite cursor by 2
+                overwritten = unsafe {
+                    OverwrittenEntry::Double(
+                        self.read_storage(self.owcurs),
+                        self.read_storage(self.owcurs + 1),
+                    )
+                };
+                self.owcurs += 2;
+            } else {
+                overwritten = unsafe { OverwrittenEntry::Single(self.read_storage(self.owcurs)) };
+                self.owcurs += 1;
+            }
+        }
+        self.write_to_storage(self.wcurs, entry);
+        self.wcurs += 1;
+        overwritten
+    }
+
+    /// Write double entry in single borrow, returning overwritten entry
+    pub fn write_double(
+        &mut self,
+        prefix: E,
+        suffix: E,
+    ) -> (OverwrittenEntry<E>, OverwrittenEntry<E>) {
+        let first_overwritten = self.write(prefix);
+        let second_overwritten = self.write(suffix);
+        (first_overwritten, second_overwritten)
+    }
+
+    /// Get value of backing storage corresponding to given index
+    #[inline]
+    pub(crate) unsafe fn read_storage(&self, curs: usize) -> E {
+        self.storage[get_cursor_index(self.storage.len(), curs, self.use_base_2_indexing)]
+            .assume_init()
+    }
+
+    #[inline]
+    fn write_to_storage(&mut self, curs: usize, entry: E) {
+        self.storage[get_cursor_index(self.storage.len(), curs, self.use_base_2_indexing)] =
+            MaybeUninit::new(entry);
+    }
+
+    #[inline]
+    pub fn read(&self, curs: usize) -> ReadEntry<E> {
```

Comment:
I don't see any tests that demonstrate the interaction of `RaceBuffer::read` or `RaceBuffer::iter` with `RaceBuffer::write` and `RaceBuffer::write_double` (the tests in lib.rs presently explore the remote-reader-style patterns). Covering the basics seems relevant here.

Reply:
This is present on `race-buffer-modifications`